Add benchmarks for forward scan and backward scan

breeswish opened the pull request:
<!--
Thank you for contributing to TiKV! Please read TiKV's [CONTRIBUTING](https://github.com/pingcap/tikv/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.
-->

## What have you changed? (mandatory)

Add benchmarks for forward scan and backward scan. The benchmark involve the gRPC and TiDB layer, which is more stable.

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Manually.

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

## Benchmark result if necessary (optional)

For each object, benchmark is run twice and the best result for each case is used to make the result more accurate.

### Mvcc Refine

PR #3458.

```
test storage::scan::bench_backward_scan_10_versions                   ... bench: 148,055,035 ns/iter (+/- 23,853,226)
test storage::scan::bench_backward_scan_1_version                     ... bench:  21,920,231 ns/iter (+/- 775,097)
test storage::scan::bench_backward_scan_1_version_long_value          ... bench:  29,081,653 ns/iter (+/- 978,382)
test storage::scan::bench_backward_scan_2_versions                    ... bench:  42,337,299 ns/iter (+/- 5,962,684)
test storage::scan::bench_backward_scan_2_versions_long_value         ... bench:  54,815,928 ns/iter (+/- 219,079)
test storage::scan::bench_backward_scan_50_versions                   ... bench: 265,795,138 ns/iter (+/- 3,291,021)
test storage::scan::bench_backward_scan_5_versions                    ... bench: 105,712,713 ns/iter (+/- 9,091,300)
test storage::scan::bench_forward_scan_10_versions                    ... bench:  43,321,426 ns/iter (+/- 619,727)
test storage::scan::bench_forward_scan_1_version                      ... bench:   6,527,396 ns/iter (+/- 589,588)
test storage::scan::bench_forward_scan_1_version_long_value           ... bench:   9,307,029 ns/iter (+/- 1,094,673)
test storage::scan::bench_forward_scan_2_versions                     ... bench:   8,857,253 ns/iter (+/- 280,964)
test storage::scan::bench_forward_scan_2_versions_long_value          ... bench:  11,236,944 ns/iter (+/- 808,247)
test storage::scan::bench_forward_scan_50_versions                    ... bench:  65,556,789 ns/iter (+/- 854,164)
test storage::scan::bench_forward_scan_5_versions                     ... bench:  22,972,381 ns/iter (+/- 3,622,228)
```

### Master

Master branch before merging #3458.

```
test storage::scan::bench_backward_scan_10_versions                   ... bench: 171,357,528 ns/iter (+/- 30,408,706)
test storage::scan::bench_backward_scan_1_version                     ... bench:  25,294,335 ns/iter (+/- 2,373,399)
test storage::scan::bench_backward_scan_1_version_long_value          ... bench:  33,686,040 ns/iter (+/- 7,004,777)
test storage::scan::bench_backward_scan_2_versions                    ... bench:  46,309,345 ns/iter (+/- 373,062)
test storage::scan::bench_backward_scan_2_versions_long_value         ... bench:  59,293,564 ns/iter (+/- 5,706,247)
test storage::scan::bench_backward_scan_50_versions                   ... bench: 477,246,421 ns/iter (+/- 99,705,009)
test storage::scan::bench_backward_scan_5_versions                    ... bench: 111,999,437 ns/iter (+/- 20,983,537)
test storage::scan::bench_forward_scan_10_versions                    ... bench:  42,795,840 ns/iter (+/- 12,345,217)
test storage::scan::bench_forward_scan_1_version                      ... bench:   9,848,621 ns/iter (+/- 666,876)
test storage::scan::bench_forward_scan_1_version_long_value           ... bench:  12,959,611 ns/iter (+/- 5,856,069)
test storage::scan::bench_forward_scan_2_versions                     ... bench:  13,050,808 ns/iter (+/- 1,133,170)
test storage::scan::bench_forward_scan_2_versions_long_value          ... bench:  15,554,615 ns/iter (+/- 531,780)
test storage::scan::bench_forward_scan_50_versions                    ... bench: 122,691,120 ns/iter (+/- 7,953,975)
test storage::scan::bench_forward_scan_5_versions                     ... bench:  27,556,550 ns/iter (+/- 2,442,810)
```

### Release 2.0

Current release-2.0 branch.

```
test storage::scan::bench_backward_scan_10_versions                  ... bench: 173,514,704 ns/iter (+/- 49,705,984)
test storage::scan::bench_backward_scan_1_version                    ... bench:  27,434,148 ns/iter (+/- 480,486)
test storage::scan::bench_backward_scan_1_version_long_value         ... bench:  35,401,608 ns/iter (+/- 425,717)
test storage::scan::bench_backward_scan_2_versions                   ... bench:  48,261,558 ns/iter (+/- 30,957,166)
test storage::scan::bench_backward_scan_2_versions_long_value        ... bench:  62,888,811 ns/iter (+/- 373,301)
test storage::scan::bench_backward_scan_50_versions                  ... bench: 464,389,978 ns/iter (+/- 58,771,614)
test storage::scan::bench_backward_scan_5_versions                   ... bench: 110,814,274 ns/iter (+/- 374,947)
test storage::scan::bench_forward_scan_10_versions                   ... bench:  45,605,278 ns/iter (+/- 3,975,140)
test storage::scan::bench_forward_scan_1_version                     ... bench:  12,121,555 ns/iter (+/- 2,317,110)
test storage::scan::bench_forward_scan_1_version_long_value          ... bench:  16,591,425 ns/iter (+/- 302,852)
test storage::scan::bench_forward_scan_2_versions                    ... bench:  14,178,577 ns/iter (+/- 3,853,202)
test storage::scan::bench_forward_scan_2_versions_long_value         ... bench:  19,896,323 ns/iter (+/- 11,932,291)
test storage::scan::bench_forward_scan_50_versions                   ... bench: 123,745,640 ns/iter (+/- 3,137,198)
test storage::scan::bench_forward_scan_5_versions                    ... bench:  29,017,557 ns/iter (+/- 3,456,863)
```

### Compare

![image](https://user-images.githubusercontent.com/1916485/44323612-3913ba80-a485-11e8-9ff1-99b4dea05908.png)

## Add a few positive/negative examples (optional)

siddontang commented:
In some cases, why `refine fixed` is better than `refine fixed + bound`?

breeswish commented:
@siddontang For forward scan 5 versions, in `refine fixed` there will be 5 next, while in `refine fixed + bound` there will be `4 next (hit seek_bound) + 1 seek`, so the later one is constantly slower.

breeswish commented:
Updated to latest result. @siddontang @zhangjinpeng1987 PTAL

zhangjinpeng1987 commented on benches/storage/scan.rs:
```diff
@@ -11,30 +11,24 @@
-        || storage::ReadPoolContext::new(pd_worker.scheduler())
-    });
-    let store = SyncStorage::new(&Default::default(), read_pool);
+fn bench_forward_scan_deleted(b: &mut Bencher) {
```
s/deleted/mvcc_deleted

zhangjinpeng1987 commented on benches/storage/scan.rs:
```diff
@@ -61,31 +55,171 @@ fn bench_tombstone_scan(b: &mut Bencher) {
+
+    println!("Preparing data...");
+
+    let value_length = if long_value {
```
I think you'd better make value_length more representative, for example 32/100/200bytes.

zhangjinpeng1987 commented on benches/storage/scan.rs:
```diff
@@ -61,31 +55,171 @@ fn bench_tombstone_scan(b: &mut Bencher) {
 }
+
+fn bench_scan(b: &mut Bencher, forward_scan: bool, versions: usize, long_value: bool) {
+    const NUMBER_KEYS: usize = 10000;
```
Please make number_keys configurable.

zhangjinpeng1987 replied:
Seems all key/value pairs are in memtable?

breeswish replied:
Seems that we need dynamic benches for different kind of keys and values.

breeswish replied:
Benchmarking 100000 keys are really slow.. even 6 hours is not enough for completing all benchmarks. Do we really have the necessity to test the difference of memtable and non-memtable access? Seems that any kind of optimizations by us won't make a difference over this.

breeswish commented:
@zhangjinpeng1987 I've changed to the following settings (and rewrote to harness benchmarks):

```rust
    for number_of_keys in &[100, 10000, 1000000] {
        for number_of_versions in &[1, 2, 5, 10, 50] {
            for key_size in &[32, 100, 200] {
                for value_size in &[5, 100, 200] {
```

Benchmark results will be updated later (it really costs time)

siddontang commented:
how does the graph generate? very cool @breeswish

breeswish commented:
@siddontang It is generated by the criterion library automatically if gnuplot is installed :)

huachaohuang commented on benches/rocksdb_writebatch.rs:
```diff
@@ -11,87 +11,40 @@
         for i in 0..batch_keys {
             let k = format!("key_round{}_key{}", r, i);
-            batch.put(k.as_bytes(), v).unwrap();
+            batch.put(black_box(k.as_bytes()), black_box(v)).unwrap();
```
Why is `black_box` necessary here?

breeswish replied:
Compiler is very smart and it will inlines constant values. In real world, these values are produced by user and it cannot be inlined, so there would be a performance difference sometimes.

breeswish commented on components/test_util/kv_generator.rs:
```diff
@@ -11,22 +11,32 @@
     key_len: usize,
     value_len: usize,
-    rng: ThreadRng,
+    rng: IsaacRng,
```
I've tested their performance difference, they are similar, so it should be ok.

breeswish replied:
The reason I choice IsaacRng here is that it accepts a seed, so that I can pass a constant seed to make benchmark values stable.

Connor1996 commented:
Please fix the conflict

breeswish commented:
/run-integration-tests

breeswish commented:
@huachaohuang @zhangjinpeng1987 Fixed conflicts. Please approve again, thanks!

breeswish commented:
@DorianZheng @zhangjinpeng1987 PTAL again, thanks!

breeswish commented:
@zhangjinpeng1987 @Hijiao @DorianZheng PTAL for this old PR, thanks! Many changes have been extracted and merged to master so that this PR becomes pretty small.

zhouqiang-cl commented:
/rebuild

zhouqiang-cl commented:
/rebuild

Hijiao commented on src/storage/mvcc/reader/mod.rs:
```diff
@@ -113,7 +113,7 @@ impl<S: Snapshot> MvccReader<S> {
     pub fn load_lock(&mut self, key: &Key) -> Result<Option<Lock>> {
         if self.scan_mode.is_some() && self.lock_cursor.is_none() {
-            let iter_opt = IterOption::new(None, None, true);
+            let iter_opt = IterOption::new(None, None, self.fill_cache);
```
@zhangjinpeng1987 here

zhangjinpeng1987 replied:
I think it is better to fill cache for lock cf. @breeswish Do you have any consideration for this which i don't notice?

breeswish replied:
It was for benchmarks that operate on fill_cache=false for more stable results. @zhangjinpeng1987 Do you think this is necessary? I can discard this change anyway.

Hoverbear commented on Makefile:
```diff
@@ -91,12 +91,15 @@ test:
 	export RUST_BACKTRACE=1 && \
 	cargo test --features "${ENABLE_FEATURES}" --all ${EXTRA_CARGO_ARGS} -- --nocapture && \
-	cargo test --features "${ENABLE_FEATURES}" --bench misc ${EXTRA_CARGO_ARGS} -- --nocapture  && \
+	cargo test --features "${ENABLE_FEATURES}" --bench misc ${EXTRA_CARGO_ARGS} -- --nocapture && \
```
You can use the same command like so: `cargo test --features "${ENABLE_FEATURES}" --bench misc --bench raftstore --bench engines ... ${EXTRA_CARGO_ARGS} -- --nocapture`

Hoverbear commented on benches/rocksdb_scan/mod.rs:
```diff
@@ -0,0 +1,225 @@
+            .unwrap();
+    }
+
+    // We should got nothing when scanning any keys.
```
```suggestion
     // We should get nothing when scanning any keys.
```

Hoverbear commented on components/bench_util/Cargo.toml:
```diff
@@ -0,0 +1,7 @@
+[package]
```
Seems a very heavyweight choice to add two functions, do you plan to expand this more later?

breeswish replied:
Looks like we have a lot of places better to use a no_cache context to make benchmarks more stable, so I extracted them into a standalone workspace.

breeswish commented:
@mapleFU You may consider re-pick this PR and propose a new one by your own.
